theseus/embodied/robot/joint.py

```diff
@@ -0,0 +1,219 @@
+        ret = angle.new_empty(angle.shape[0], 3, 4)
+        ret[:, :, :3] = self.origin[:, :, :3] @ rot
+        ret[:, :, 3] = self.origin[:, :, 3]
+        return ret
+
+
+class RevoluteJoint(_RevoluteJointImpl):
+    def __init__(
+        self,
+        angle_axis: torch.Tensor,
+        name: str,
+        parent: int = -1,
+        child: int = -1,
+        origin: Optional[torch.Tensor] = None,
+        dtype: Optional[torch.dtype] = None,
+    ):
+        if angle_axis.ndim == 1:
+            angle_axis = angle_axis.view(-1, 1)
+
+        if angle_axis.ndim != 2 or angle_axis.shape != (3, 1):
+            raise ValueError("The angle axis must be a 3-D vector.")
+
+        super().__init__(name, parent, child, origin, dtype)
+
+        if angle_axis.dtype != self.dtype:
+            raise ValueError(f"The dtype of angle_axis should be {self.dtype}.")
+
+        self._axis[3:] = angle_axis
+
+    def _rotation_impl(self, angle: torch.Tensor) -> torch.Tensor:
+        if angle.ndim == 1:
+            angle = angle.view(-1, 1)
+        if angle.ndim != 2 or angle.shape[1] != 1:
+            raise ValueError("The joint angle must be a vector.")
+        return so3.exp(angle @ self.axis[3:].view(1, -1))
+
+
+class RevoluteJointX(_RevoluteJointImpl):
+    def __init__(
+        self,
+        name: str,
+        parent: int = -1,
+        child: int = -1,
+        origin: Optional[torch.Tensor] = None,
+        dtype: Optional[torch.dtype] = None,
+    ):
+        super().__init__(name, parent, child, origin, dtype)
+        self._axis[3] = 1
+
+    def _rotation_impl(self, angle: torch.Tensor) -> torch.Tensor:
+        if angle.ndim == 2 and angle.shape[1] == 1:
+            angle = angle.view(-1)
+        if angle.ndim != 1:
+            raise ValueError("The joint angle must be a vector.")
+        rot = angle.new_zeros(angle.shape[0], 3, 3)
+        rot[:, 0, 0] = 1
+        rot[:, 1, 1] = angle.cos()
+        rot[:, 2, 1] = angle.sin()
+        rot[:, 1, 2] = -rot[:, 2, 1]
+        rot[:, 2, 2] = rot[:, 1, 1]
+        return rot
+
+
+class RevoluteJointY(_RevoluteJointImpl):
+    def __init__(
+        self,
+        name: str,
+        parent: int = -1,
+        child: int = -1,
+        origin: Optional[torch.Tensor] = None,
+        dtype: Optional[torch.dtype] = None,
+    ):
+        super().__init__(name, parent, child, origin, dtype)
+        self._axis[4] = 1
+
+    def _rotation_impl(self, angle: torch.Tensor) -> torch.Tensor:
+        if angle.ndim == 2 and angle.shape[1] == 1:
+            angle = angle.view(-1)
+        if angle.ndim != 1:
+            raise ValueError("The joint angle must be a vector.")
+        rot = angle.new_zeros(angle.shape[0], 3, 3)
+        rot[:, 1, 1] = 1
+        rot[:, 0, 0] = angle.cos()
+        rot[:, 0, 2] = angle.sin()
+        rot[:, 2, 0] = -rot[:, 0, 2]
+        rot[:, 2, 2] = rot[:, 0, 0]
+        return rot
+
+
+class RevoluteJointZ(_RevoluteJointImpl):
```

Comment:
Wondering if the following alternative is better/worse in any way: only one `RevoluteJoint` class that has a full, x, y, or z type as a property (this would get rid of the Impl intermediate as well).

Reply:
Yeah, this crossed my mind as well. I think I'm OK with either.

Reply:
I am okay with either as well. We should consider tradeoffs like ease of use and potential overheads.

Reply:
Now there are two classes for `RevoluteJoint`, i.e., `RevoluteJointFull` and `RevoluteJointXYZ`